### DISCUSSION TITLE
Is creating a delegate better than an interface invocation?

### DISCUSSION THREAD

**Discussion by paulomorgado** [dotnet/runtime#53862] [Q&A] [answered]
`GenericArraySortHelper<T>`.Sort receives an `IComparer<T>` and creates a `IComparision<T>` delegate on every invocation:

[`ArraySortHelper<T>.IntrospectiveSort(keys, comparer.Compare);`](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs#L307)

Is this better than an invocation of `IComparer<T>.Compare`?

**Comment by stephentoub** [comment, marked as answer]
https://github.com/dotnet/coreclr/pull/8504

**Reply by paulomorgado** [reply]
I've created [this benchmark](https://github.com/paulomorgado/Benchmarks/tree/main/ComparerComparison), but it's not very conclusive. Or is it?

**Reply by dellamonica** [reply]
That is a nice question.

Perhaps in your benchmark the JIT is able to know exactly the IComparer<T> type in your Sort method, so it devirtualizes the call, which makes it slightly faster than going through the comparison delegate.

On the other hand, if the JIT can't know the type of the comparer with certainty, I'm guessing it will be slower than the delegate call.

**Reply by gfoidl** [reply]
> the JIT can't know the type of the comparer with certainty

This where dynamic profile guided optimization (D-PGO) will hook in with tiered compilation. On a high-level: in Tier-0 the code gets instrumented, i.e. it is recorded which types for the comparer are present, then in Tier-1 a hot-path with de-virtualized code is emitted in companion to the cold-path consisting of the interface-dispatch.

> I'm guessing

With more and more cool optimization tricks coming to JIT the only way to know is to measure. Fortunately / Unfortunately (depending on the point of view) plain guessing or experience may become wrong now.

**Reply by dellamonica** [reply]
Yes, PGO is indeed a game changer.

Currently I have some high-perf code for which several implementations of a given interface are called in a fairly uniform distribution. I'm using this delegate wrapping strategy, but if PGO manages to beat it without the extra complication, it would be certainly welcome.
